pallets/tendermint-light-client/src/lib.rs

```diff
@@ -0,0 +1,337 @@
+
+        /// Maximal number of finalized headers to keep in the storage, last-in first-out
+        #[pallet::constant]
+        type HeadersToKeep: Get<u32>;
+
+        /// Maximal number of block validators in Tendermint
+        #[pallet::constant]
+        type MaxVotesCount: Get<u32>;
+
+        /// time provider type, used to gauge whether blocks are within the trusting period
+        type TimeProvider: UnixTime;
+    }
+
+    #[pallet::pallet]
+    #[pallet::storage_version(STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Pallet is halted
+        LightClientHalted,
+        /// Pallet operations are resumed        
+        LightClientResumed,
+        /// Light client is initialized
+        LightClientInitialized,
+        /// New block has been verified and imported into storage \[relayer_address, imported_block_hash\]
+        ImportedLightBlock(T::AccountId, TendermintBlockHash),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Light client has not been initialized        
+        NotInitialized,
+        /// Light client has already been initialized
+        AlreadyInitialized,
+        /// Light client is currently halted
+        Halted,
+        /// The minimum voting power threshold is not reached, the block cannot be trusted yet
+        NotEnoughTrust,
+        /// Verification failed, the block is invalid        
+        InvalidBlock,
+        /// Initial block is invalid
+        InvalidInitialBlock,
+        /// General error during client operations
+        Other,
+    }
+
+    // NOTE for storage:
+    // - header storage should be a ring buffer (i.e. we keep n last headers, ordered by the insertion time)
+    // - we keep a pointer to the last finalized header (to avoid deserializing the whole buffer)
+    // - insertion moves the pointer and updates the buffer
+    // or?
+    // https://substrate.recipes/ringbuffer.html
+
+    /// Hash of the last imported header from the bridged chain
+    #[pallet::storage]
+    #[pallet::getter(fn get_last_imported_block_hash)]
+    pub type LastImportedBlockHash<T: Config> = StorageValue<_, TendermintBlockHash, ValueQuery>;
+
+    /// Imported hashes "ordered" by their insertion time
+    /// Client keeps HeadersToKeep number of these at any time    
+    #[pallet::storage]
+    #[pallet::getter(fn get_imported_hash)]
+    pub type ImportedHashes<T: Config> = StorageMap<_, Identity, u32, TendermintBlockHash>;
+
+    /// Current ring buffer position
+    #[pallet::storage]
+    #[pallet::getter(fn get_imported_hashes_pointer)]
+    pub(super) type ImportedHashesPointer<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+    /// Bridged chain Headers which have been imported by the client
+    /// Client keeps HeadersToKeep number of these at any time
+    #[pallet::storage]
+    #[pallet::getter(fn get_imported_block)]
```

Comment:
```suggestion
    #[pallet::getter(fn get_imported_blocks)]
```

Should be plural, correct?

Reply:
no - the storage is a map, so the auto-generated getter returns one block (value) by it's hash (key).

Reply:
Ah, ok. Thanks.